broadcaster/_backends/pulsar.py

```diff
@@ -14,64 +15,80 @@ def __init__(self, url: str):
         self._port = parsed_url.port or 6650
         self._service_url = f"pulsar://{self._host}:{self._port}"
         self._client = None
-        self._producer = None
-        self._consumer = None
+        self._producers = {}
+        self._consumers = {}
+        self._subscribed_channels = set()
 
     async def connect(self) -> None:
         try:
-            logging.info("Connecting to brokers")
+            logger.info("Connecting to Pulsar brokers")
             self._client = await anyio.to_thread.run_sync(
                 lambda: pulsar.Client(self._service_url)
             )
-            self._producer = await anyio.to_thread.run_sync(
-                lambda: self._client.create_producer("broadcast")
-            )
-            self._consumer = await anyio.to_thread.run_sync(
-                lambda: self._client.subscribe(
-                    "broadcast",
-                    subscription_name="broadcast_subscription",
-                    consumer_type=pulsar.ConsumerType.Shared,
-                )
-            )
-            logging.info("Successfully connected to brokers")
+            logger.info("Successfully connected to Pulsar brokers")
         except Exception as e:
-            logging.error(e)
+            logger.error(f"Error connecting to Pulsar: {e}")
             raise e
 
     async def disconnect(self) -> None:
-        if self._producer:
-            await anyio.to_thread.run_sync(self._producer.close)
-        if self._consumer:
-            await anyio.to_thread.run_sync(self._consumer.close)
+        for producer in self._producers.values():
+            await anyio.to_thread.run_sync(producer.close)
+        for consumer in self._consumers.values():
+            await anyio.to_thread.run_sync(consumer.close)
         if self._client:
             await anyio.to_thread.run_sync(self._client.close)
 
     async def subscribe(self, channel: str) -> None:
-        # In this implementation, we're using a single topic 'broadcast'
-        # So we don't need to do anything here
-        pass
+        if channel not in self._subscribed_channels:
+            self._subscribed_channels.add(channel)
+            consumer = await anyio.to_thread.run_sync(
+                lambda: self._client.subscribe(
+                    channel,
+                    subscription_name=f"broadcast_subscription_{channel}",
+                    consumer_type=pulsar.ConsumerType.Shared,
+                )
+            )
+            self._consumers[channel] = consumer
+            logger.info(f"Subscribed to channel: {channel}")
 
     async def unsubscribe(self, channel: str) -> None:
-        # Similarly, we don't need to do anything here
-        pass
+        if channel in self._subscribed_channels:
+            self._subscribed_channels.remove(channel)
+            consumer = self._consumers.pop(channel, None)
+            if consumer:
```

Comment:
Better to first close the consumer then remove it from the dict. If an error is raised attempting to close, we will not have access to the channel anymore